src/Conversion/ONNXToKrnl/Tensor/GatherElements.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===--------- GatherElements.cpp - Lowering GatherElements Op ----------===//
+//
+// Copyright 2022 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX GatherElements Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.hpp"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+struct ONNXGatherElementsOpLowering : public ConversionPattern {
+  ONNXGatherElementsOpLowering(TypeConverter &typeConverter, MLIRContext *ctx)
+      : ConversionPattern(
+            typeConverter, ONNXGatherElementsOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXGatherElementsOpAdaptor operandAdaptor(operands);
+    ONNXGatherElementsOp gatherElementsOp = cast<ONNXGatherElementsOp>(op);
+    Location loc = op->getLoc();
+
+    ONNXGatherElementsOpShapeHelper shapeHelper(&gatherElementsOp, &rewriter,
+        krnl::getDenseElementAttributeFromKrnlValue,
+        krnl::loadDenseElementArrayValueAtIndex);
+    auto shapecomputed = shapeHelper.computeShape(operandAdaptor);
+    assert(succeeded(shapecomputed) && "Could not compute output shape");
+
+    // Insert an allocation and deallocation for the result of this operation.
+    MemRefType outputMemRefType = convertToMemRefType(*op->result_type_begin());
+    Value output = insertAllocAndDeallocSimple(
+        rewriter, op, outputMemRefType, loc, shapeHelper.dimsForOutput());
+
+    // Operands and attributes.
+    Value data = operandAdaptor.data();
+    Value indices = operandAdaptor.indices();
+    int64_t axis = gatherElementsOp.axis();
+    int64_t dataRank = data.getType().cast<ShapedType>().getRank();
```

Comment:
`ShapedType` can be unranked, e.g. `TensorType<*xtype>`, and `getRank()` would fail. Maybe, `MemRefType` is better.

Reply:
Using MemrefType.